dev/source/docs/debugging.rst

```diff
@@ -6,6 +6,69 @@ Debugging
 
 This section is for topics related to debugging ArduPilot source code.
 
+Debug Output
+------------
+
+There are three ways to easily get debug output from your code - so called "printf debugging".
+
+MAVLink messages
+................
+
+These messages will appear on Ground Control Stations (and will also appear in DataFlash logs).
+
+Most places in the code you can insert a single statement to send a text message to the GCS, using the GCS singleton:
+
+::
+
+   gcs().send_text(MAV_SEVERITY_WARNING, "Best number ever: %f", 42.4242424242);
+
+
+DataFlash Log messages
+......................
+
+The DataFlash singleton can be used to put messages to the DataFlash log:
+
+::
+
+   DataFlash_Class::instance()->Log_Write_MessageF("u-blox %d HW: %s SW: %s",
```

Comment:
need to use new AP::Write() API

Reply:
http://ardupilot.org/dev/docs/code-overview-adding-a-new-log-message.html